src/System.Net.Http/src/System/Net/Http/HttpContent.cs

```diff
@@ -115,30 +112,19 @@ public Task<string> ReadAsStringAsync()
                 // then check for a byte-order-mark (BOM) in the data to figure out the encoding.
                 if (encoding == null)
                 {
-                    byte[] preamble;
-                    foreach (Encoding testEncoding in s_encodingsWithBom)
-                    {
-                        preamble = testEncoding.GetPreamble();
-                        if (ByteArrayHasPrefix(data, dataLength, preamble))
-                        {
-                            encoding = testEncoding;
-                            bomLength = preamble.Length;
-                            break;
-                        }
-                    }
+                    TryDetectEncoding(data, dataLength, ref encoding, ref bomLength);
                 }
 
-                // Use the default encoding if we couldn't detect one.
-                encoding = encoding ?? DefaultStringEncoding;
-
-                // BOM characters may be present even if a charset was specified.
-                if (bomLength == -1)
+                if (encoding == null)
                 {
-                    byte[] preamble = encoding.GetPreamble();
-                    if (ByteArrayHasPrefix(data, dataLength, preamble))
-                        bomLength = preamble.Length;
-                    else
-                        bomLength = 0;
+                    // Use the default encoding if we couldn't detect one.
+                    encoding = DefaultStringEncoding;
+                    bomLength = 0;
+                }
+                else if (bomLength == -1)
```

Comment:
> Do you have a preference?

My preference would be the Try version, as it makes it more clear to me at least what policy is being applied. If you did switch to the DetectEncoding version, my preference would be for it to return the encoding rather than using a ref, and to use out rather ref for the bomLength. Just my personal preference.